### DISCUSSION TITLE
asgi.App should not be a sub-class of falcon.App

### DISCUSSION THREAD

**Discussion by davetapley** [falconry/falcon#2210] [Ideas]
This is motivated by https://github.com/falconry/falcon/pull/2209#issuecomment-1922220895, TLDR:

Because an `asgi.App` _is_ (in the [LSP](https://en.wikipedia.org/wiki/Liskov_substitution_principle) sense) a `falcon.App`,
any type for a `falcon.App` must be 'polluted' with types which only make sense in the ASGI context. E.g.:

The `simulate_request` methods in `testing.client` must return `Result | StreamedResult`,
even though it's not possible to get a `StreamedResult` with WSGI (aka `falcon.App`).

That's awkward because `StreamedResult` doesn't have the **very useful** `text` and `json` methods.

---

I would like to propose moving the shared functionality out of `falcon.App` and in to a mix-in class.
Then methods like ⬇️ could be extended with a [`TypeGuard`](https://peps.python.org/pep-0647/#typeguard-type).
https://github.com/falconry/falcon/blob/4910dd73ecd1b9c8cf6cae045b26ad432fa56128/falcon/testing/client.py#L2177

and it will be possible to use [`@overload`](https://docs.python.org/3/library/typing.html#typing.overload) to correctly narrow to just `Result`.

I _suspect_ there are other places in the code which would benefit from this (e.g. all the ASGI only websocket code).

Thoughts?

**Comment by CaselIT** [comment]
Hi,

Please take a look at https://github.com/falconry/falcon/pull/2211

**Comment by CaselIT** [comment]
> any type for a falcon.App must be 'polluted' with types which only make sense in the ASGI context. E.g.:

I don't follow, you can surely create a type guard to accept only a subclass. What you can't do is the inverse, you can't create a type guard for the superclass

**Reply by davetapley** [reply]
Correct! I should have been more explicit:

The 'problem' (with the current implementation) is that _if_ a method needs to accept a ASGI app, _and_ return a different type (if iff app is ASGI), then the method will need to return both types (because you could always replace the WSGI app with the an ASGI app, because LSP).

However: That's not a problem for #2211 because (as I now realize, thanks to your changes 🙏🏻 ) `def simulate_request` would never return `StreamedResult` anyway (since there's no way to set `_stream_result` and `_one_shot`). Knowing that I'm happy to leave this for now.

**Reply by CaselIT** [reply]
In a sense yes, but I don't think it's a popular use case to pass the App around to different methods.
In any case it seems that overloads can be used to change the behavior in case it's needed, so it may be good enough for now?

**Comment by vytas7** [comment]
I don't think the mentioned points are insurmountable issues, but in general -- yes, unfortunately `asgi.App` subclassing [wsgi] `App` violates the [LSP](https://en.wikipedia.org/wiki/Liskov_substitution_principle) for methods that become `async` or take coroutines as arguments in the superclass, such as `App.add_sink()`, `App.__call__()`, etc.

That would be a major undertaking to change the hierarchy in Falcon 5.0, so far I'm leaning towards keeping the things as is, since the typecheckers don't seem to complain in the downstream (i.e., end user) apps if we exempt these errors on our side.

**Reply by davetapley** [reply]
To be sure:
> since the typecheckers don't seem to complain in the downstream (i.e., end user) apps

is excluding the initial `Result | StreamedResult` problem from "downstream (i.e., end user) apps", or has that been fixed elsewhere?

**Reply by vytas7** [reply]
Hey @davetapley, no, sorry, I was focusing more on the general use of `App` vs `asgi.App`, not the test client and its methods!
I will take a look at the `Result` | `StreamedResult` problem.

**Reply by vytas7** [reply]
Hi again @davetapley, I believe thanks to the referenced PRs by @CaselIT, this is largely resolved for the ordinary end user, although the internal typing of the framework is somewhat tricky.

Consider the following file (that passes `pytest`):
```python
import falcon
import falcon.asgi
import falcon.testing


class HelloResource:
    def on_get(self, req: falcon.Request, resp: falcon.Response) -> None:
        resp.media = {'message': 'Hello'}


class AsyncResource:
    async def on_get(self, req: falcon.Request, resp: falcon.Response) -> None:
        resp.content_type = falcon.MEDIA_TEXT
        resp.text = 'Hello, async!\n'


def simulate_hello(app: falcon.App | falcon.asgi.App) -> falcon.testing.Result:
    return falcon.testing.simulate_get(app, '/hello')


def simulate_with_client(client: falcon.testing.TestClient) -> falcon.testing.Result:
    return client.simulate_get('/hello')


def test_sync() -> None:
    app = falcon.App()
    app.add_route('/hello', HelloResource())

    result = simulate_hello(app)
    assert result.json == {'message': 'Hello'}


def test_async() -> None:
    app = falcon.asgi.App()
    app.add_route('/hello', AsyncResource())

    result = simulate_hello(app)
    assert result.text == 'Hello, async!\n'


def test_using_client() -> None:
    client = falcon.testing.TestClient(falcon.App())
    result = simulate_with_client(client)
    assert result.status_code == 404
    assert result.json == {'title': '404 Not Found'}
```

Assuming we have CPython 3.13 or 3.14 environment (older CPython would require explicitly [parametrizing `App` with response and request types](https://falcon.readthedocs.io/api/typing.html#app-types) to pass strict typing):

```bash
$ mypy --strict test.py 
Success: no issues found in 1 source file
```

```bash
$ pyright test.py 
0 errors, 0 warnings, 0 informations
```

Normally, you just simulate directly, but I explicitly went through `Result` in different ways to make sure it stays correctly typed at all times.

So everything should work... Unless I misunderstood your concerns (once again)?

**Reply by davetapley** [reply]
@vytas7 perfect, thanks! And thank you @CaselIT 🙏🏻
